### PR TITLE
8308609: java/lang/ScopedValue/StressStackOverflow.java fails with "-XX:-VMContinuations"

### PR DESCRIPTION
**Pull request by theRealAph**
Move `runWith()` from `VirtualThread` to `BaseVirtualThread`.

`BoundVirtualThread` does not use `runWith()` to run its task, so when a VM error occurs it can not recover scoped values.

Moving `runWith()` into the common subclass of both `VirtualThread` and `BoundVirtualThread` fixes the problem.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308609](https://bugs.openjdk.org/browse/JDK-8308609): java/lang/ScopedValue/StressStackOverflow.java fails with "-XX:-VMContinuations" (**Bug** - P4)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [b2663471](https://git.openjdk.org/jdk/pull/14399/files/b2663471519b031139fc9224d0fa1cc2955712e8)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14399/head:pull/14399` \
`$ git checkout pull/14399`

Update a local copy of the PR: \
`$ git checkout pull/14399` \
`$ git pull https://git.openjdk.org/jdk.git pull/14399/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14399`

View PR using the GUI difftool: \
`$ git pr show -t 14399`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14399.diff">https://git.openjdk.org/jdk/pull/14399.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14399#issuecomment-1584890249)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/BaseVirtualThread.java [hunk 1]
```diff
@@ -63,5 +68,13 @@ sealed abstract class BaseVirtualThread extends Thread
...
      */
     abstract void unpark();
+
+    @Hidden
+    @ForceInline
+    protected void runWith(Object bindings, Runnable op) {
+        ensureMaterializedForStackWalk(bindings);
+        op.run();
+        Reference.reachabilityFence(bindings);
+    }
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/lang/BaseVirtualThread.java:78]: BaseVirtualThread <: Thread.  Would it be simpler, and avoid duplicate, to just change Thread.runWith to be a final method with default access. That would be accessibility to VirtualThread and BoundVirtualThread without duplication.

**Comment by offamitkumar** [comment]
Hi Andrew, 

This test failure is no more reproducible on s390x now, would you consider undo-ing problemlisting done by [JDK-8303499](https://bugs.openjdk.org/browse/JDK-8303499), I'll close the parent task [JDK-8303498](https://bugs.openjdk.org/browse/JDK-8303498)

File: src/java.base/share/classes/java/lang/BaseVirtualThread.java (hunk 1 above)
Comment (theRealAph) [reply src/java.base/share/classes/java/lang/BaseVirtualThread.java:78]: Sure, that makes sense.

**Comment by AlanBateman** [comment]
One suggestion for the test is to have it re-run with -XX:-VMContinuations on platforms/ports that do do have VM continuations support. Some of the tests for virtual threads do the same thing to ensure this operating mode is tested. This should do it:

```
--- a/test/jdk/java/lang/ScopedValue/StressStackOverflow.java
+++ b/test/jdk/java/lang/ScopedValue/StressStackOverflow.java
@@ -21,8 +21,8 @@
  * questions.
  */
 
-/**
- * @test
+/*
+ * @test id=default
  * @summary StressStackOverflow the recovery path for ScopedValue
  * @enablePreview
  * @run main/othervm/timeout=300 -XX:-TieredCompilation StressStackOverflow
@@ -30,6 +30,13 @@
  * @run main/othervm/timeout=300 StressStackOverflow
  */
 
+/*
+ * @test id=no-vmcontinuations
+ * @requires vm.continuations
+ * @enablePreview
+ * @run main/othervm/timeout=300 -XX:+UnlockExperimentalVMOptions -XX:-VMContinuations StressStackOverflow
+ */
+
```

File: test/jdk/ProblemList-Virtual.txt [hunk 2]
```diff
@@ -44,8 +44,6 @@ javax/management/remote/mandatory/loading/MissingClassTest.java 8145413 windows-
 
 javax/management/remote/mandatory/loading/RMIDownloadTest.java 8308366 windows-x64
 
-java/lang/ScopedValue/StressStackOverflow.java 8309646 generic-all
-
```
Comment (AlanBateman) [inline test/jdk/ProblemList-Virtual.txt:48]: That's the issue for the JTREG_TEST_THREAD_FACTORY=Virtual runs. I think we need to leave it excluded until that issue is diagnosed/fixed. 

However, in jdk/test/ProblemList.txt, it is excluded on linux-s390x as that platform doesn't seem to have, or use, VM continuations and so is using the fallback/alternative implementation. I think this entry can be removed now.

Comment (theRealAph) [reply test/jdk/ProblemList-Virtual.txt:48]: I think it's fixed now.

Two problems:
1. Test was not guaranteed to make forward progress. Fixed with a 3 minute timeout.
2. Test was very slow in some conditions. Likewise.

**Review by AlanBateman — APPROVED** [review]
The implementation changes look good.

The update to test looks okay too although I would like to see if we can do better in SVC and STS when there is a SO.  In passing, I think all usages of tlr should be replaced with ThreadLocalRandom.current().

**Comment by theRealAph** [comment]
> The implementation changes look good.
> 
> The update to test looks okay too although I would like to see if we can do better in SVC and STS when there is a SO.

That would be nice, but it's something of a can of worms. 

> In passing, I think all usages of tlr should be replaced with ThreadLocalRandom.current().

Err, sure -- but why? I guess there may be something I don't understand here.

**Review by AlanBateman — APPROVED** [review]
I'd probably go further and drop the TLR parameter from the choose method too but what you have is fine, let's go with that.

**Comment by AlanBateman** [comment]
I didn't spot this before it was integrated but test/hotspot/jtreg/serviceability/jvmti/thread/GetStackTrace/getstacktr03/libgetstacktr03.cpp is sensitive to the stack trace so its like of expected frames will be take account of runWith no longer be overridden in the sub-class.

**Comment by dcubed-ojdk** [comment]
@theRealAph -
So the GHA results for this PR show that serviceability/jvmti/thread/GetStackTrace/getstacktr03/getstacktr03
failed. Why wasn't this addressed prior to integration? We saw 5 instances of this failure mode in
every Mach5 Tier1 and 20 instances in every Mach5 Tier3 after this PR was integrated. This was not
an intermittent failure.

I've applied the fix that @AlanBateman proposed in [JDK-8310211](https://bugs.openjdk.org/browse/JDK-8310211)
and that appears to have solved the issue for Mach5 Tier1. I'll have to wait for the
fix to reach Mach5 Tier3 before I can declare that one is also good.

**Pull request closed without merging**